src/RingGrids/general.jl

```diff
@@ -0,0 +1,398 @@
+# if no nlat_half provided calculate it
+function (::Type{Grid})(data::AbstractArray) where {Grid<:AbstractGridArray}
+    npoints2D = size(data, 1)                   # from 1st dim of data
+    nlat_half = get_nlat_half(Grid, npoints2D)  # get nlat_half of Grid
+    return Grid(data, nlat_half)
+end
+
+for f in (:zeros, :ones, :rand, :randn)
+    @eval begin
+        # general version with ArrayType(zeros(...)) conversion
+        function Base.$f(
+            ::Type{Grid},
+            nlat_half::Integer,
+            k::Integer...,
+        ) where {Grid<:AbstractGridArray{T, N, ArrayType}} where {T, N, ArrayType}
+            return Grid(ArrayType($f(T, get_npoints2D(Grid, nlat_half), k...)), nlat_half)
+        end
+
+        # CPU version with zeros(T, ...) producing Array
+        function Base.$f(
+            ::Type{Grid},
+            nlat_half::Integer,
+            k::Integer...,
+        ) where {Grid<:AbstractGridArray{T}} where T
+            return Grid($f(T, get_npoints2D(Grid, nlat_half), k...), nlat_half)
+        end
+
+        # use Float64 if no type provided
+        function Base.$f(
+            ::Type{Grid},
+            nlat_half::Integer,
+            k::Integer...,
+        ) where {Grid<:AbstractGridArray}
+            return $f(Grid{Float64}, nlat_half, k...)
+        end
+    end
+end
+
+# zero element of an AbstractGridArray instance grid by creating new zero(grid.data)
+Base.zero(grid::Grid) where {Grid<:AbstractGridArray} =
+    nonparametric_type(Grid)(zero(grid.data), grid.nlat_half, grid.rings)
+
+# similar data but everything else identical
+function Base.similar(grid::Grid) where {Grid<:AbstractGridArray}
+    return nonparametric_type(Grid)(similar(grid.data), grid.nlat_half, grid.rings)
+end
+
+# data with new type T but everything else identical
+function Base.similar(grid::Grid, ::Type{T}) where {Grid<:AbstractGridArray, T}
+    return nonparametric_type(Grid)(similar(grid.data, T), grid.nlat_half, grid.rings)
+end
+
+# data with same type T but new size
+function Base.similar(
+    grid::Grid,
+    nlat_half::Integer,
+    k::Integer...
+) where {Grid<:AbstractGridArray{T, N, ArrayType}} where {T, N, ArrayType}
+    similar_data = similar(grid.data, get_npoints2D(Grid, nlat_half), k...)
+    return nonparametric_type(Grid)(similar_data, nlat_half)
+end
+
+# data with new type T and new size
+function Base.similar(
+    grid::Grid,
+    ::Type{Tnew},
+    nlat_half::Integer,
+    k::Integer...
+) where {Grid<:AbstractGridArray, Tnew}
+    similar_data = similar(grid.data, Tnew, get_npoints2D(Grid, nlat_half), k...)
+    return nonparametric_type(Grid)(similar_data, nlat_half)
+end
+
+# general version with ArrayType{T, N}(undef, ...) generator
+function (::Type{Grid})(
+    ::UndefInitializer,
+    nlat_half::Integer,
+    k::Integer...,
+) where {Grid<:AbstractGridArray{T, N, ArrayType}} where {T, N, ArrayType}
+    return Grid(ArrayType{T, N}(undef, get_npoints2D(Grid, nlat_half), k...), nlat_half)
```

Comment:
I am not sure this works. 

`ArrayType` is already something like `Array{T,1}`, so you can't do `Array{T,1}{T,N}`. 
Is this function tested with JLArray?

Maybe we should really just define 

```Julia
nonparametric_type(::Array{T,N}) where {T,N} = Array
nonparametric_type(::CuArray{T,N}) where {T,N} = CuArray

# and in the tests, but not the model:
import SpeedyWeather.nonparametric_type
SpeedyWeather.nonparametric_type(::JLArray{T,N}) where {T,N} = JLArray
```
This could then also be used by the broadcast

Reply:
Yes, `CuArray` is exported by CUDA.jl 

`JLArray` really has zero productive use. An extension would work as well, but they are really only used for tests, so both is fine. 

Yep, the parameters aren't necessary.

Reply:
`nonparametric_type` was added for `Array` and `CuArray` and as extension also for `JLArray`